turbopack/crates/turbo-tasks-fs/src/watcher.rs

```diff
@@ -252,11 +300,49 @@ impl DiskWatcher {
         let mut batched_new_paths = FxHashSet::default();
 
         'outer: loop {
-            let mut event = rx.recv().or(Err(TryRecvError::Disconnected));
+            let mut event_result = rx.recv().or(Err(TryRecvError::Disconnected));
+            // this inner loop batches events using `try_recv`
             loop {
-                match event {
-                    Ok(Ok(notify::Event { kind, paths, .. })) => {
-                        let paths: Vec<PathBuf> = paths
+                match event_result {
+                    Ok(Ok(event)) => {
+                        // TODO: We might benefit from some user-facing diagnostics if it rescans
+                        // occur frequently (i.e. more than X times in Y minutes)
+                        //
+                        // You can test rescans on Linux by reducing the inotify queue to something
+                        // really small:
+                        //
+                        // ```
+                        // echo 3 | sudo tee /proc/sys/fs/inotify/max_queued_events
+                        // ```
+                        if event.need_rescan() {
+                            let _lock = inner.invalidation_lock.blocking_write();
+
+                            #[cfg(not(any(target_os = "macos", target_os = "windows")))]
+                            {
+                                // we can't narrow this down to a smaller set of paths: Rescan
+                                // events (at least when tested on Linux) come with no `paths`.
+                                self.restore_all_watching(inner.root_path());
+                                batched_new_paths.clear();
+                            }
+
+                            if report_invalidation_reason {
+                                inner.invalidate_with_reason();
```

Comment:
The reason should be "rescan" instead of "invalidate" since this invalidation is triggered by a watch event and not a manual invalidation.